pkg/workspace/assert.go

```diff
@@ -0,0 +1,167 @@
+	return false
+}
+
+// Visit if `node` is annotated with `@assert/validate` (AnnotationAssertValidate).
+// Checks, validates, and runs the assertions, any violations from running the assertions are collected and stored.
+//
+// This visitor returns and error if any assertion is not well-formed,
+// otherwise, returns nil.
+func (a *assertChecker) Visit(node yamlmeta.Node) error {
+	nodeAnnotations := template.NewAnnotations(node)
+	if nodeAnnotations.Has(AnnotationAssertValidate) {
+		validations, syntaxErr := newValidationsFromAssertValidateAnnotation(nodeAnnotations[AnnotationAssertValidate], node, a.thread)
+		if syntaxErr != nil {
+			return syntaxErr
+		}
+		for _, v := range validations {
+			err := v.Validate(node)
+			if err != nil {
+				a.violations = append(a.violations, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func newValidationsFromAssertValidateAnnotation(annotation template.NodeAnnotation, n yamlmeta.Node, thread *starlark.Thread) ([]Validation, error) {
+	var validations []Validation
+	validationPosition := createAssertAnnotationPosition(annotation.Position, n)
+
+	if len(annotation.Kwargs) != 0 {
+		return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found keyword argument (by %v)", AnnotationAssertValidate, validationPosition.AsCompactString())
+	}
+	if len(annotation.Args) == 0 {
+		return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found no arguments (by %v)", AnnotationAssertValidate, validationPosition.AsCompactString())
+	}
+	for _, arg := range annotation.Args {
+		validationTuple, ok := arg.(starlark.Tuple)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have validation 2-tuple as argument(s), but found: %v (by %v)", AnnotationAssertValidate, arg.String(), validationPosition.AsCompactString())
+		}
+		if len(validationTuple) != 2 {
+			return nil, fmt.Errorf("Expected @%v 2-tuple, but found tuple with length %v (by %v)", AnnotationAssertValidate, len(validationTuple), validationPosition.AsCompactString())
+		}
+		message, ok := validationTuple[0].(starlark.String)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have descriptive string as the first item in the 2-tuple, but found type: %v (by %v)", AnnotationAssertValidate, validationTuple[0].Type(), validationPosition.AsCompactString())
+		}
+		lambda, ok := validationTuple[1].(starlark.Callable)
+		if !ok {
+			return nil, fmt.Errorf("Expected @%v to have a validation function as the second item in the 2-tuple, but found type: %v (by %v)", AnnotationAssertValidate, validationTuple[1].Type(), validationPosition.AsCompactString())
+		}
+		validations = append(validations, Validation{
+			thread:   thread,
+			msg:      message.String(),
+			f:        lambda,
+			position: validationPosition,
+		})
+	}
+
+	return validations, nil
+}
+
+// Validation is an assertion from an @assert/validate annotation.
+// One @assert/validate annotation can have multiple Validations.
+type Validation struct {
```

Comment:
I think what we have here is a `Rule`???

From: https://github.com/vmware-tanzu/carvel/pull/331/files#diff-a81faf6d33e82e22eee1083ee484a23eda8a928b046116a2cb516e58086984b8R48-R53

> - **rule** — the combination of: 
>     - a textual description of what constitutes a valid value;
>     - a function that asserts the rule against an actual value;
>     - a violation message template, used to report invalid values.
>     - **custom rule** — a rule based on a function (either one written by a user or provided in a `ytt` module).
>     - **named rule** — a rule provided with `ytt` expressed through a keyword argument to an `@schema/validation` or `@assert/validate`.

vs.

> - **validation** — the binding of one or more "rule"s to a "node".

ℹ️  what words we use is absolutely up for debate... 👍🏻   What I'm pushing for is [ubiquitous language](https://thedomaindrivendesign.io/developing-the-ubiquitous-language/).

Reply:
Thank you for this 🙌  I didn't catch that I had some of these terms incorrect/swapped.